core/domain/user_services.py

```diff
@@ -2287,3 +2291,272 @@ def get_dashboard_stats(user_id):
+            current exploration in sequential order.
+        checkpoints_in_older_exploration: list(str). The checkpoints
+            of older exploration in sequential order.
+        most_distant_reached_checkpoint_state_name_in_older_exploration: str.
+            The state name of the most distant reached checkpoint in the older
+            exploration.
+
+    Returns:
+        str or None. The most distant checkpoint in current exploration or
+        None if most distant reached checkpoint of older exploration is not
+        present in current exploration.
+    """
+    # Index of the most_distant_reached_checkpoint in the older exploration.
+    mdrc_index = (
+        checkpoints_in_older_exploration.index(
+            most_distant_reached_checkpoint_state_name_in_older_exploration))
+
+    # Loop through checkpoints of furthest_reached_exploration backwards until
+    # a checkpoint is found that exists in current_exploration too.
+    while mdrc_index >= 0:
+        if checkpoints_in_older_exploration[mdrc_index] in checkpoints_in_current_exploration: # pylint: disable=line-too-long
+            return checkpoints_in_older_exploration[mdrc_index]
+        mdrc_index = mdrc_index - 1
+
+    return None
+
+
+def update_learner_checkpoint_progress(
+    user_id, exploration_id, state_name, exp_version
+):
+    """Sets the furthest reached and most recently reached checkpoint in
+    an exploration by the user.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+        state_name: str. The state name of the most recently reached checkpoint.
+        exp_version: int. The exploration version of the most recently reached
+            checkpoint.
+    """
+
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    if exploration_user_model is None:
+        exploration_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exploration_id)
+
+    current_exploration = exp_fetchers.get_exploration_by_id(
+        exploration_id, True, exp_version)
+
+    # If the exploration is being visited the first time.
+    if exploration_user_model.furthest_reached_checkpoint_state_name is None:
+        exploration_user_model.furthest_reached_checkpoint_exp_version = (
+            exp_version)
+        exploration_user_model.furthest_reached_checkpoint_state_name = (
+            state_name)
+    else:
+        furthest_reached_checkpoint_exploration = (
+            exp_fetchers.get_exploration_by_id(
+                exploration_id,
+                True,
+                exploration_user_model.furthest_reached_checkpoint_exp_version
+            )
+        )
+        checkpoints_in_current_exploration = _get_checkpoints_in_order(
+            current_exploration.init_state_name, current_exploration.states)
+        checkpoints_in_furthest_reached_exploration = _get_checkpoints_in_order(
```

Comment:
I would call this checkpoints_in_older_exploration

Also could you shortcut some of this logic if the exploration_user_model.furthest_reached_checkpoint_exp_version matches the current exp version?

Reply:
Done!